src/hypothesis/searchstrategy/types.py

```diff
@@ -0,0 +1,260 @@
+# coding=utf-8
+#
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis-python
+#
+# Most of this work is copyright (C) 2013-2017 David R. MacIver
+# (david@drmaciver.com), but it contains contributions by others. See
+# CONTRIBUTING.rst for a full list of people who may hold copyright, and
+# consult the git log if you need to determine who owns an individual
+# contribution.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at http://mozilla.org/MPL/2.0/.
+#
+# END HEADER
+
+from __future__ import division, print_function, absolute_import
+
+import io
+import sys
+import functools
+import collections
+
+import hypothesis.strategies as st
+from hypothesis.errors import ResolutionFailed
+from hypothesis.internal.compat import text_type, integer_types
+
+_global_type_to_strategy_lookup = {}
+
+
+def required_args(spec, args=(), kwargs=()):
+    """Return a set of required args in this argspec."""
+    return set(spec.args + spec.kwonlyargs) \
+        - set(spec.args[len(spec.args) - len(spec.defaults or ()):]) \
+        - set(spec.kwonlydefaults or ()) - set(args) - set(kwargs)
+
+
+@st.cacheable
+def type_strategy_mapping():
+    """Return a dict mapping from types to corresponding search strategies.
+
+    Most resolutions will terminate here or in the special handling for
+    generics from the typing module.
+
+    """
+    import uuid
+    import decimal
+    import datetime as dt
+    import fractions
+
+    known_type_strats = {
+        # Types with core Hypothesis strategies
+        type(None): st.none(),
+        bool: st.booleans(),
+        float: st.floats(),
+        complex: st.complex_numbers(),
+        fractions.Fraction: st.fractions(),
+        decimal.Decimal: st.decimals(),
+        text_type: st.text(),
+        bytes: st.binary(),
+        dt.datetime: st.datetimes(),
+        dt.date: st.dates(),
+        dt.time: st.times(),
+        dt.timedelta: st.timedeltas(),
+        uuid.UUID: st.uuids(),
+        # Built-in types
+        type: st.sampled_from([type(None), bool, int, str, list, set, dict]),
+        type(Ellipsis): st.just(Ellipsis),
+        type(NotImplemented): st.just(NotImplemented),
+        bytearray: st.binary().map(bytearray),
+        memoryview: st.binary().map(memoryview),
+        # Pull requests with more types welcome!
+    }
+    for t in integer_types:
+        known_type_strats[t] = st.integers()
+    # build empty collections, as only generics know their contents
+    known_type_strats.update({
+        t: st.builds(t) for t in (tuple, list, set, frozenset, dict)
+    })
+    try:
```

Comment:
I feel like this would be more cleanly handled by having the extras modules be responsible for defining the type mappings and having some code that tries to load each extras module.

Also, should there be support for Django models? the `models` strategy seems like a natural fit with this feature.